uniffi_bindgen/src/interface/types/dispatch.rs

```diff
@@ -0,0 +1,196 @@
+        match $self {
+            Type::UInt8 => SimpleTypeHandler::UInt8.$fn_name($($param),*),
+            Type::Int8 => SimpleTypeHandler::Int8.$fn_name($($param),*),
+            Type::UInt16 => SimpleTypeHandler::UInt16.$fn_name($($param),*),
+            Type::Int16 => SimpleTypeHandler::Int16.$fn_name($($param),*),
+            Type::UInt32 => SimpleTypeHandler::UInt32.$fn_name($($param),*),
+            Type::Int32 => SimpleTypeHandler::Int32.$fn_name($($param),*),
+            Type::UInt64 => SimpleTypeHandler::UInt64.$fn_name($($param),*),
+            Type::Int64 => SimpleTypeHandler::Int64.$fn_name($($param),*),
+            Type::Float32 => SimpleTypeHandler::Float32.$fn_name($($param),*),
+            Type::Float64 => SimpleTypeHandler::Float64.$fn_name($($param),*),
+            Type::Boolean => SimpleTypeHandler::Boolean.$fn_name($($param),*),
+            Type::String => SimpleTypeHandler::String.$fn_name($($param),*),
+            Type::Timestamp => TimestampTypeHandler.$fn_name($($param),*),
+            Type::Duration => DurationTypeHandler.$fn_name($($param),*),
+            Type::Object(name) => ObjectTypeHandler { name }.$fn_name($($param),*),
+            Type::Record(name) => RecordTypeHandler { name }.$fn_name($($param),*),
+            Type::Enum(name) => EnumTypeHandler { name }.$fn_name($($param),*),
+            Type::Error(name) => ErrorTypeHandler { name }.$fn_name($($param),*),
+            Type::CallbackInterface(name) => CallbackInterfaceTypeHandler {
+                name,
+            }.$fn_name($($param),*),
+            Type::Optional(inner) => OptionalTypeHandler { inner }.$fn_name($($param),*),
+            Type::Sequence(inner) => SequenceTypeHandler { inner }.$fn_name($($param),*),
+            Type::Map(inner) => MapTypeHandler { inner }.$fn_name($($param),*),
+            Type::External { name, crate_name } => ExternalTypeHandler {
+                name,
+                crate_name,
+            }.$fn_name($($param),*),
+            Type::Wrapped { name, prim } => WrapperTypeHandler {
+                name,
+                wrapped: prim.as_ref(),
+            }.$fn_name($($param),*),
+        }
+    }
+);
+
+// The `type_dispatch!` trait wrapper.  F
+
+macro_rules! type_dispatch {
+    (
+        $(#[$meta:meta])*
+        $vis:vis trait $name:ident $tt:tt
+    ) => {
+        $(#[$meta])*
+        $vis trait $name $tt
+
+        type_dispatch_impl_for_type!($tt, $name);
+        type_dispatch_impl_for_related!(crate::interface::Record, $tt, $name);
+        type_dispatch_impl_for_related!(crate::interface::Enum, $tt, $name);
+        type_dispatch_impl_for_related!(crate::interface::CallbackInterface, $tt, $name);
+        type_dispatch_impl_for_related!(crate::interface::Error, $tt, $name);
+        type_dispatch_impl_for_related!(crate::interface::Object, $tt, $name);
+    };
+    (
+        $(#[$meta:meta])*
+        $vis:vis trait $name:ident : $super:ident $tt:tt
```

Comment:
It took me a little bit of staring to understand that the only difference between these two rules was the `super` trait. Please consider adding a comment to that effect :-)

Alternately, would something like `$( : $super:ident)?` work here to handle both in a single rule? I don't have much experience with optional matching in macro rules so I'm not sure what's possible.

Reply:
I think I tried that but couldn't get it to work.  I think the issue there is that an `ident` is also a  `tt`, so it makes the rules ambiguous.